### PR TITLE
New feature: OpenAPI generator for kotlin

### PR DESCRIPTION
**Pull request by altro3**

### CODE REVIEW

**Comment by altro3** [comment]
@graemerocher don't know how to fix binary comatibility problems. Also I found bug with micronaut-core for kotlin - optional QueryValues are always required. And validation don't work for controller parameters

File: openapi-generator/src/main/resources/templates/kotlin-micronaut/client/auth/Authorization.mustache [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+{{#generatedAnnotation}}
+{{>common/generatedAnnotation}}
+{{/generatedAnnotation}}
+@MustBeDocumented
+@Retention(RUNTIME)
+@Target(FUNCTION)
+@Bindable
+@Repeatable
+annotation class Authorization(
```
Comment (graemerocher) [inline openapi-generator/src/main/resources/templates/kotlin-micronaut/client/auth/Authorization.mustache:20]: in general I am wondering why we don't upstream more of this annotation/filter logic

**Comment by altro3** [comment]
@graemerocher about you question: I don’t see any problems with generating code in Kotlin, but implementing interfaces in Java. I fixed what you didn't like - now all tests for Kotlin are written in Kotlin.

But! Thanks to the fact that I initially wrote tests in Java, I was able to find validation bugs if I mixed Java and Kotlin in one project. I’m not sure if this is popular, but I don’t understand why some of the methods work correctly, while others give errors in this case

File: openapi-generator/src/main/resources/templates/kotlin-micronaut/client/auth/Authorization.mustache (hunk 1 above)
Comment (altro3) [reply openapi-generator/src/main/resources/templates/kotlin-micronaut/client/auth/Authorization.mustache:20]: I don't know, that's what was in the java generator. I just converted this to kotlin

File: openapi-generator/src/main/resources/templates/kotlin-micronaut/client/auth/configuration/HttpBasicAuthConfiguration.mustache [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import io.micronaut.http.MutableHttpRequest
+{{#generatedAnnotation}}
+import {{javaxPackage}}.annotation.Generated
+{{/generatedAnnotation}}
+
+{{#generatedAnnotation}}
+{{>common/generatedAnnotation}}
+{{/generatedAnnotation}}
+@EachProperty("security.basic-auth")
+open class HttpBasicAuthConfiguration @ConfigurationInject constructor(
```
[the remark this answers is not in the document]
Comment (altro3) [reply openapi-generator/src/main/resources/templates/kotlin-micronaut/client/auth/configuration/HttpBasicAuthConfiguration.mustache]: fixed

File: openapi-generator/src/main/resources/templates/kotlin-micronaut/common/test/model_test.mustache [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * Model tests for {{classname}}
+ */
+@MicronautTest
+class {{classname}}Test {
+
+    {{#models}}
+    {{#model}}
+    {{^vendorExtensions.x-is-one-of-interface}}
+    {{^isEnum}}
+    val model: {{classname}}? = null
```
[the remark this answers is not in the document]
Comment (altro3) [reply openapi-generator/src/main/resources/templates/kotlin-micronaut/common/test/model_test.mustache:19]: fixed

File: test-suite-kotlin-server-generator/src/main/java/io/micronaut/openapi/test/api/ParametersController.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import io.micronaut.openapi.test.model.ColorEnum;
+import io.micronaut.openapi.test.model.SendDatesResponse;
+import io.micronaut.openapi.test.model.SendPrimitivesResponse;
+
+import org.jetbrains.annotations.NotNull;
+
+import reactor.core.publisher.Mono;
+
+@Controller
+public class ParametersController implements ParametersApi {
```
[the remark this answers is not in the document]
Comment (altro3) [reply test-suite-kotlin-server-generator/src/main/java/io/micronaut/openapi/test/api/ParametersController.java]: look to my comment. I converted code from java to kotlin

File: openapi-generator/src/main/resources/templates/kotlin-micronaut/common/test/model_test.mustache (hunk 3 above)
[the remark this answers is not in the document]
Comment (altro3) [reply openapi-generator/src/main/resources/templates/kotlin-micronaut/common/test/model_test.mustache]: Can' t do it

Initialization exception

**Review by graemerocher — APPROVED** [review]

**Comment by andriy-dmytruk** [comment]
I feel like there is too much duplication between Kotlin & Java. 
I want to take a look at whether we can avoid that and will try to get back to you some day this week.

**Comment by altro3** [comment]
@andriy-dmytruk I did this on purpose for a specific reason:

I don’t want the generator code for Kotlin to affect an already existing generator for Java (the same is done in the openapi generator itself). Yes, there are duplicates, but you can be sure that if you correct the logic for Kotlin, you will not change the logic for Java and vice versa.

Unfortunately, no solution will be perfect. Excessive unification always leads to more complex code and potential problems in the future.

**Review by andriy-dmytruk — APPROVED** [review]

**Comment by altro3** [comment]
@andriy-dmytruk @graemerocher Could you merge it?

**Comment by graemerocher** [comment]
@altro3 Thanks for the contribution again. I think you can probably do a release, you don't need to wait for 4.2

**Pull request merged**